File: server-rails-hybrid/app/policies/project_policy.rb
# frozen_string_literal: true

class ProjectPolicy < Rhino::ResourcePolicy
  def permitted_attributes_for_show(user)
    if has_role?(user, "owner") || has_role?(user, "admin")
      ["*"]
    elsif has_role?(user, "manager")
      %w[id title description status budget starts_at ends_at]
    elsif has_role?(user, "member") || has_role?(user, "viewer")
      %w[id title description status starts_at ends_at]
    else
      []
    end
  end

  def hidden_attributes_for_show(user)
    if has_role?(user, "manager")
      %w[internal_notes]
    elsif has_role?(user, "member") || has_role?(user, "viewer")
      %w[budget internal_notes]
    else
      []
    end
  end

  def permitted_attributes_for_create(user)
    if has_role?(user, "owner") || has_role?(user, "admin")
      %w[title description status budget internal_notes starts_at ends_at]
    elsif has_role?(user, "manager")
      %w[title description status starts_at ends_at]
    else
      []
    end
  end

  def permitted_attributes_for_update(user)
    if has_role?(user, "owner") || has_role?(user, "admin")
      %w[title description status budget internal_notes starts_at ends_at]
    elsif has_role?(user, "manager")
      %w[title description status starts_at ends_at]
    else
      []
    end
  end
end

File: server-nestjs-single/src/policies/ProjectPolicy.ts
import { ResourcePolicy } from '@rhino-dev/rhino-nestjs';

/**
 * Project policy — SINGLE-TENANT (role-free).
 *
 * No roles in this variant: action authorization is driven by the user's global
 * `permissions` (User.permissions JSON) checked by ResourcePolicyGuard via
 * `{slug}.{action}`. Row-level ownership is enforced separately by ProjectScope
 * (userId == current user), so attribute permissions are simply "the owner sees
 * and edits everything they own".
 *
 * `userId` is permitted on create because StampProjectOwner stamps it from the
 * authenticated user before validation — it is never client-trusted.
 */
export class ProjectPolicy extends ResourcePolicy {
  override resourceSlug = 'projects';

  override permittedAttributesForShow(): string[] {
    return ['*'];
  }
  override hiddenAttributesForShow(): string[] {
    return [];
  }
  override permittedAttributesForCreate(): string[] {
    return ['userId', 'title', 'description', 'status', 'budget', 'internalNotes', 'startsAt', 'endsAt'];
  }
  override permittedAttributesForUpdate(): string[] {
    return ['title', 'description', 'status', 'budget', 'internalNotes', 'startsAt', 'endsAt'];
  }
}
